diff --git a/src/lj_record.c b/src/lj_record.c
index 1dd22dac..e930eba6 100644
--- a/src/lj_record.c
+++ b/src/lj_record.c
@@ -443,6 +448,19 @@ static void rec_for_check(jit_State *J, IRType t, int dir,
   }
 }
 
+/* Convert IR to the target number type. Used for loop indexes narrowing. */
+static TRef irconv_numtype(jit_State *J, TRef tr, IRType t)
+{
+  if (t == IRT_INT) {
+    if (!tref_isinteger(tr))
+      return emitir(IRTGI(IR_CONV), tr, IRCONV_INT_NUM|IRCONV_CHECK);
+  } else {
+    if (!tref_isnum(tr))
+      return emitir(IRTN(IR_CONV), tr, IRCONV_NUM_INT);
+  }
+  return tr;
+}
+
 /* Record a FORL instruction. */
 static void rec_for_loop(jit_State *J, const BCIns *fori, ScEvEntry *scev,
 			 int init)
@@ -454,8 +472,8 @@ static void rec_for_loop(jit_State *J, const BCIns *fori, ScEvEntry *scev,
 	     (init || LJ_DUALNUM) ? lj_opt_narrow_forl(J, tv) : IRT_NUM;
   int mode = IRSLOAD_INHERIT +
     ((!LJ_DUALNUM || tvisint(tv) == (t == IRT_INT)) ? IRSLOAD_READONLY : 0);
-  TRef stop = fori_arg(J, fori, ra+FORL_STOP, t, mode);
-  TRef step = fori_arg(J, fori, ra+FORL_STEP, t, mode);
+  TRef stop = irconv_numtype(J, fori_arg(J, fori, ra+FORL_STOP, t, mode), t);
+  TRef step = irconv_numtype(J, fori_arg(J, fori, ra+FORL_STEP, t, mode), t);
   int tc, dir = rec_for_direction(&tv[FORL_STEP]);
   lj_assertJ(bc_op(*fori) == BC_FORI || bc_op(*fori) == BC_JFORI,
 	     "bad bytecode %d instead of FORI/JFORI", bc_op(*fori));
@@ -517,13 +535,7 @@ static LoopEvent rec_for(jit_State *J, const BCIns *fori, int isforl)
       lj_assertJ(tref_isnumber_str(tr[i]), "bad FORI argument type");
       if (tref_isstr(tr[i]))
 	tr[i] = emitir(IRTG(IR_STRTO, IRT_NUM), tr[i], 0);
-      if (t == IRT_INT) {
-	if (!tref_isinteger(tr[i]))
-	  tr[i] = emitir(IRTGI(IR_CONV), tr[i], IRCONV_INT_NUM|IRCONV_CHECK);
-      } else {
-	if (!tref_isnum(tr[i]))
-	  tr[i] = emitir(IRTN(IR_CONV), tr[i], IRCONV_NUM_INT);
-      }
+      tr[i] = irconv_numtype(J, tr[i], t);
     }
     tr[FORL_EXT] = tr[FORL_IDX];
     stop = tr[FORL_STOP];
